Basic authentication

Using the `hapi-auth-basic` plugin, add a new `/private` endpoint protected by a username and password that returns a simple HTML page with a welcome message. The welcome message should include the logged in username. Do not use any templates (views) at this stage. For username/password validation, create a static `lib/users.json` file (we will replace this in the near future) that contains a few sample users and passwords.

The new endpoint will be contained in its own plugin, similar to how we set up `/version`, and will require bringing test coverage back to 100%.

Due: 4/27

Second question.

I rewrote index.js to this so it's register more then 1 plugin. 

~~~

var Hapi = require('hapi');
var Version = require('./version');
var Authentication = require('./authentication')

// Declare internals

var internals = {};


exports.init = function (port, next) {

    var server = new Hapi.Server();
    server.connection({ port: port });

   server.register([
       {
           register: require('Version'),
           options: {} 
           },
       {
           register: require('Authentication'),
           options: {} 
       }
       ], function (err) {
       
       if (err) {
            return next(err);
        }

        server.start(function (err) {

            return next(err, server);
        });
    });

~~~

but on testing and running I see this error message : 

Error: Cannot find module 'Version' 

What did I have done wrong ?

--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 test:
 	@node node_modules/lab/bin/lab -a code
 test-cov:
-	@node node_modules/lab/bin/lab -a code -t 100 -L
+	@node node_modules/lab/bin/lab -a code -t 100 -Lv
 test-cov-html:
 	@node node_modules/lab/bin/lab -a code -r html -o coverage.html
 

--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 test:
 	@node node_modules/lab/bin/lab -a code
 test-cov:
-	@node node_modules/lab/bin/lab -a code -t 100 -Lv
+	@node node_modules/lab/bin/lab -a code -t 100 -Lv --lint-errors-threshold 0 --lint-warnings-threshold 0
 test-cov-html:
 	@node node_modules/lab/bin/lab -a code -r html -o coverage.html
 

--- a/lib/index.js
+++ b/lib/index.js
@@ -2,6 +2,7 @@
 
 var Hapi = require('hapi');
 var Version = require('./version');
+var Private = require('./private');
 
 
 // Declare internals
@@ -14,7 +15,7 @@ exports.init = function (port, next) {
     var server = new Hapi.Server();
     server.connection({ port: port });
 
-    server.register(Version, function (err) {
+    server.register([Version, Private], function (err) {
 
         if (err) {
             return next(err);

--- a/lib/private.js
+++ b/lib/private.js
@@ -1,0 +1,49 @@
+// Load modules
+
+var Basic = require('hapi-auth-basic');
+var Users = require('./users.json');
+
+// Declare internals
+
+var internals = {};
+
+internals.validateFunc = function (username, password, callback) {
+
+    var user = Users[username];
+    if (!user || user.password !== password) {
+        return callback(null, false);
+    }
+
+    return callback(null, true, user);
+};
+
+exports.register = function (server, options, next) {
+
+    server.register(Basic, function (err) {
+
+        if (err) {
+            return next(err);
+        }
+
+        server.auth.strategy('simple', 'basic', { validateFunc: internals.validateFunc });
+        server.route({
+            method: 'GET',
+            path: '/private',
+            config: {
+                auth: 'simple',
+                description: 'Returns a greeting message to the authenticated user',
+                handler: function (request, reply) {
+
+                    var html = '<div>Hello ' + request.auth.credentials.username + '</div>';
+                    return reply(html);
+                }
+            }
+        });
+
+        return next();
+    });
+};
+
+exports.register.attributes = {
+    name: 'Private'
+};

--- a/lib/users.json
+++ b/lib/users.json
@@ -1,0 +1,10 @@
+{
+	"Foo": {
+		"username": "Foo",
+		"password": "mysupersecuredpassword"
+	},
+	"Bar": {
+		"username": "Bar",
+		"password": "mysupersecuredpassword2"
+	}
+}

--- a/package.json
+++ b/package.json
@@ -23,7 +23,7 @@
     "hoek": "2.x.x"
   },
   "scripts": {
-    "test": "node node_modules/lab/bin/lab -a code -t 100 -Lv",
+    "test": "node node_modules/lab/bin/lab -a code -t 100 -Lv --lint-errors-threshold 0 --lint-warnings-threshold 0",
     "start": "node lib/start.js"
   },
   "devDependencies": {

--- a/package.json
+++ b/package.json
@@ -19,10 +19,11 @@
   "homepage": "https://github.com/hueniverse/hueniversity",
   "dependencies": {
     "hapi": "8.x.x",
+    "hapi-auth-basic": "2.x.x",
     "hoek": "2.x.x"
   },
   "scripts": {
-    "test": "node node_modules/lab/bin/lab -a code -t 100 -L",
+    "test": "node node_modules/lab/bin/lab -a code -t 100 -Lv",
     "start": "node lib/start.js"
   },
   "devDependencies": {

--- a/test/index.js
+++ b/test/index.js
@@ -10,7 +10,6 @@ var Version = require('../lib/version');
 // Test shortcuts
 
 var lab = exports.lab = Lab.script();
-var describe = lab.experiment;
 var expect = Code.expect;
 var it = lab.test;
 
@@ -50,7 +49,7 @@ it('handles register plugin errors', { parallel: false }, function (done) {
         name: 'fake version'
     };
 
-    Hueniversity.init(0, function (err, server) {
+    Hueniversity.init(0, function (err) {
 
         expect(err).to.exist();
         expect(err.message).to.equal('register version failed');

--- a/test/private.js
+++ b/test/private.js
@@ -75,6 +75,7 @@ describe('/private', function () {
         var orig = Basic.register;
         Basic.register = function (plugin, options, next) {
 
+            Basic.register = orig;
             return next(new Error('fail'));
         };
         Basic.register.attributes = {
@@ -82,8 +83,6 @@ describe('/private', function () {
         };
 
         Hueniversity.init(0, function (err) {
-
-            Basic.register = orig;
 
             expect(err).to.exist();
 

--- a/test/private.js
+++ b/test/private.js
@@ -1,0 +1,98 @@
+// Load modules
+
+var Code = require('code');
+var Lab = require('lab');
+var Hueniversity = require('../lib');
+var Users = require('../lib/users.json');
+var Basic = require('hapi-auth-basic');
+
+// Declare internals
+
+var internals = {};
+
+
+// Test shortcuts
+
+var lab = exports.lab = Lab.script();
+var describe = lab.experiment;
+var expect = Code.expect;
+var it = lab.test;
+
+
+describe('/private', function () {
+
+    it('returns a greeting for the authenticated user', function (done) {
+
+        Hueniversity.init(0, function (err, server) {
+
+            expect(err).to.not.exist();
+
+            var request = { method: 'GET', url: '/private', headers: { authorization: internals.header(Users.Foo.username, Users.Foo.password) } };
+            server.inject(request, function (res) {
+
+                expect(res.statusCode, 'Status code').to.equal(200);
+                expect(res.result, 'result').to.equal('<div>Hello Foo</div>');
+
+                server.stop(done);
+            });
+        });
+    });
+
+    it('returns error on wrong password', function (done) {
+
+        Hueniversity.init(0, function (err, server) {
+
+            expect(err).to.not.exist();
+
+            var request = { method: 'GET', url: '/private', headers: { authorization: internals.header(Users.Foo.username, '') } };
+            server.inject(request, function (res) {
+
+                expect(res.statusCode, 'Status code').to.equal(401);
+
+                server.stop(done);
+            });
+        });
+    });
+
+    it('returns error on failed auth', function (done) {
+
+        Hueniversity.init(0, function (err, server) {
+
+            expect(err).to.not.exist();
+
+            var request = { method: 'GET', url: '/private', headers: { authorization: internals.header('I do not exist', '') } };
+            server.inject(request, function (res) {
+
+                expect(res.statusCode, 'Status code').to.equal(401);
+
+                server.stop(done);
+            });
+        });
+    });
+
+    it('returns error on failed registering of auth', { parallel: false }, function (done) {
+
+        var orig = Basic.register;
+        Basic.register = function (plugin, options, next) {
+
+            return next(new Error('fail'));
+        };
+        Basic.register.attributes = {
+            name: 'fake hapi-auth-basic'
+        };
+
+        Hueniversity.init(0, function (err) {
+
+            Basic.register = orig;
+
+            expect(err).to.exist();
+
+            done();
+        });
+    });
+});
+
+internals.header = function (username, password) {
+
+    return 'Basic ' + (new Buffer(username + ':' + password, 'utf8')).toString('base64');
+};

--- a/test/version.js
+++ b/test/version.js
@@ -3,7 +3,7 @@
 var Code = require('code');
 var Lab = require('lab');
 var Pkg = require('../package.json');
-var Server = require('../lib');
+var Hueniversity = require('../lib');
 
 
 // Test shortcuts
@@ -18,7 +18,7 @@ describe('/version', function () {
 
     it('returns the version from package.json', function (done) {
 
-        Server.init(0, function (err, server) {
+        Hueniversity.init(0, function (err, server) {
 
             expect(err).to.not.exist();
 

